File: packages/aap-android/src/main/kotlin/dev/aap/sdk/AAPRelay.kt
package dev.aap.sdk

import kotlinx.coroutines.*
import kotlinx.serialization.json.*
import okhttp3.*
import java.util.concurrent.TimeUnit

/**
 * Real-time session relay over WebSocket.
 *
 * Mirrors the JS `RelayTransport` and Swift `AAPRelay` exactly.
 * The server forwards frames opaquely — end-to-end encryption is
 * maintained; the relay never sees plaintext.
 *
 * Usage:
 * ```kotlin
 * val relay = AAPRelay(handle = "alice.android", mailboxUrl = "wss://mailbox.aap.dev")
 * relay.connect(did = identity.did)
 * relay.onReceive { frame -> ... }
 * relay.send(frameBytes)
 * ```
 */
class AAPRelay(
    private val handle:     String,
    private val mailboxUrl: String
) {
    sealed class Event {
        data class Waiting(val handle: String) : Event()
        data class Ready(val handle: String)   : Event()
        data class Closed(val by: String)      : Event()
        data class Error(val message: String)  : Event()
    }

    private var ws:             WebSocket?  = null
    private var onMessageCb:    ((ByteArray) -> Unit)? = null
    private var onEventCb:      ((Event) -> Unit)?     = null
    private var connectDeferred = CompletableDeferred<Unit>()
    private var _isReady        = false

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(0,  TimeUnit.MILLISECONDS)  // no timeout on WebSocket reads
        .build()

    /** Connect to the relay. Suspends until SESSION_WAITING or SESSION_READY. */
    suspend fun connect(did: String, signature: String? = null) {
        val wsUrl  = mailboxUrl
            .replace("https://", "wss://")
            .replace("http://",  "ws://")
            .trimEnd('/')
        val encoded = java.net.URLEncoder.encode(handle, "UTF-8")
        val url     = "$wsUrl/v1/session/$encoded/ws"
        val auth    = if (signature != null) "DID $did $signature" else "DID $did"

        val request = Request.Builder()
            .url(url)
            .header("Authorization", auth)
            .build()

        connectDeferred = CompletableDeferred()

        ws = client.newWebSocket(request, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                handleFrame(text.toByteArray())
            }

            override fun onMessage(webSocket: WebSocket, bytes: okio.ByteString) {
                handleFrame(bytes.toByteArray())
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                connectDeferred.completeExceptionally(AAPError.HandshakeFailed(t.message ?: "WebSocket failure"))
                onEventCb?.invoke(Event.Error(t.message ?: "unknown"))
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                _isReady = false
                onEventCb?.invoke(Event.Closed(reason))
            }
        })

        withTimeout(15_000) { connectDeferred.await() }
    }

    private fun handleFrame(data: ByteArray) {
        val text = data.toString(Charsets.UTF_8)
        try {
            val obj  = Json.parseToJsonElement(text).jsonObject
            val type = obj["type"]?.jsonPrimitive?.content
            when (type) {
                "SESSION_WAITING" -> {
                    _isReady = true
                    connectDeferred.complete(Unit)
                    onEventCb?.invoke(Event.Waiting(handle))
                    return
                }
                "SESSION_READY" -> {
                    _isReady = true
                    connectDeferred.complete(Unit)
                    onEventCb?.invoke(Event.Ready(handle))
                    return
                }
                "SESSION_CLOSED" -> {
                    _isReady = false
                    onEventCb?.invoke(Event.Closed(obj["by"]?.jsonPrimitive?.content ?: ""))
                    return
                }
                "ERROR" -> {
                    val msg = obj["error"]?.jsonPrimitive?.content ?: "unknown"
                    connectDeferred.completeExceptionally(AAPError.HandshakeFailed(msg))
                    onEventCb?.invoke(Event.Error(msg))
                    return
                }
            }
        } catch (_: Exception) { /* not a control frame */ }
        onMessageCb?.invoke(data)
    }

    fun send(data: ByteArray) {
        ws?.send(okio.ByteString.of(*data))
    }

    fun send(text: String) {
        ws?.send(text)
    }

    fun onReceive(handler: (ByteArray) -> Unit)  { onMessageCb = handler }
    fun onEvent(handler:   (Event) -> Unit)       { onEventCb   = handler }

    fun close() {
        _isReady = false
        ws?.close(1000, "done")
    }

    val isReady: Boolean get() = _isReady
}

File: packages/aap-swift/Sources/AAPKit/AAPRelay.swift
import Foundation

/// Real-time session relay over WebSocket.
///
/// Mirrors the JS `RelayTransport` class exactly.
/// Connect two agents across any platform — the relay server forwards
/// frames without decrypting them, preserving end-to-end encryption.
///
/// Usage:
/// ```swift
/// let relay = AAPRelay(handle: "alice.dev", mailboxURL: URL(string: "wss://mailbox.aap.dev")!)
/// try await relay.connect(did: identity.did)
/// relay.onReceive { frame in ... }
/// relay.send(data: frameData)
/// ```
@MainActor
public final class AAPRelay: NSObject, @unchecked Sendable {

    // MARK: - State

    public enum Event: Sendable {
        case waiting(handle: String)
        case ready(handle: String)
        case closed(by: String)
        case error(String)
    }

    private var task:       URLSessionWebSocketTask?
    private let handle:     String
    private let mailboxURL: URL
    private var did:        String = ""

    private var onMessageHandler: ((Data) -> Void)?
    private var onEventHandler:   ((Event) -> Void)?
    private var connectContinuation: CheckedContinuation<Void, Error>?
    private var _isReady = false

    // MARK: - Init

    public init(handle: String, mailboxURL: URL) {
        self.handle     = handle
        self.mailboxURL = mailboxURL
    }

    // MARK: - Connect

    /// Connect to the relay. Resolves when the session is WAITING or READY.
    public func connect(did: String, signature: String? = nil) async throws {
        self.did = did

        // Build WebSocket URL: replace http(s) with ws(s)
        var components        = URLComponents(url: mailboxURL, resolvingAgainstBaseURL: false)!
        components.scheme     = mailboxURL.scheme == "https" ? "wss" : "ws"
        components.path       = "/v1/session/\(handle.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? handle)/ws"
        guard let wsURL = components.url else {
            throw AAPError.handshakeFailed("Invalid relay URL")
        }

        var request = URLRequest(url: wsURL)
        let auth    = signature != nil ? "DID \(did) \(signature!)" : "DID \(did)"
        request.setValue(auth, forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 10

        task = URLSession.shared.webSocketTask(with: request)
        task?.resume()

        // Wait for SESSION_WAITING or SESSION_READY
        try await withCheckedThrowingContinuation { (cont: CheckedContinuation<Void, Error>) in
            self.connectContinuation = cont
            self.startReceiveLoop()
        }
    }

    // MARK: - Send

    public func send(data: Data) {
        guard _isReady else { return }
        task?.send(.data(data)) { _ in /* ignore send errors */ }
    }

    public func send(string: String) {
        guard _isReady else { return }
        task?.send(.string(string)) { _ in }
    }

    // MARK: - Handlers

    public func onReceive(_ handler: @escaping (Data) -> Void) {
        onMessageHandler = handler
    }

    public func onEvent(_ handler: @escaping (Event) -> Void) {
        onEventHandler = handler
    }

    // MARK: - Disconnect

    public func close() {
        _isReady = false
        task?.cancel(with: .normalClosure, reason: nil)
    }

    public var isReady: Bool { _isReady }

    // MARK: - Receive loop

    private func startReceiveLoop() {
        task?.receive { [weak self] result in
            guard let self else { return }
            switch result {
            case .failure(let error):
                Task { @MainActor in
                    self.connectContinuation?.resume(throwing: error)
                    self.connectContinuation = nil
                }
            case .success(let message):
                let data: Data
                switch message {
                case .data(let d):   data = d
                case .string(let s): data = Data(s.utf8)
                @unknown default:    data = Data()
                }
                Task { @MainActor in
                    self.handleFrame(data)
                    self.startReceiveLoop()
                }
            }
        }
    }

    private func handleFrame(_ data: Data) {
        // Try to parse as a control frame
        if let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let type = obj["type"] as? String {
            switch type {
            case "SESSION_WAITING":
                _isReady = true
                connectContinuation?.resume()
                connectContinuation = nil
                onEventHandler?(.waiting(handle: handle))
                return
            case "SESSION_READY":
                _isReady = true
                connectContinuation?.resume()
                connectContinuation = nil
                onEventHandler?(.ready(handle: handle))
                return
            case "SESSION_CLOSED":
                _isReady = false
                onEventHandler?(.closed(by: obj["by"] as? String ?? ""))
                return
            case "ERROR":
                let msg = obj["error"] as? String ?? "unknown"
                connectContinuation?.resume(throwing: AAPError.handshakeFailed(msg))
                connectContinuation = nil
                onEventHandler?(.error(msg))
                return
            default:
                break
            }
        }
        // Data frame — pass to consumer
        onMessageHandler?(data)
    }
}
